### ISSUE TITLE
Azure CLI needs to be able to delete pipeline runs

### ISSUE THREAD

**Issue by swenkeratmicrosoft** [Azure/azure-cli#17968]
**Is your feature request related to a problem? Please describe.**
During testing, I sometimes create a _lot_ of runs and artifacts.  I can easily exceed how much storage I have available in a single _day_.  

**Describe the solution you'd like**
The following command should delete a run including all artifacts.
az pipelines runs delete --id [int]

**Describe alternatives you've considered**
Right now, the only way to delete a run is to manually go to the website and delete it individually.  Doing this manually for dozens of runs is painful.

**Comment by yonzhan** [comment]
route to service team

**Comment by gautelo** [comment]
When making a mistake generating a lot of failed runs and want to clean up, this is exactly the kind of feature you look for.

The proposed solution would be sufficient, though I would like it even more if you could do this in bulk by passing the same flags that the list command does.

Typical workflow would be to run az pipelines list and tweak flags until you've filtered down to what you want, and then change list to delete and run again.

**Comment by odelljl** [comment]
I agree this would be a very useful feature.  Another use case would be to delete all pipeline runs that are not run against the main or master branch. I often accumulate pipeline runs when testing a new or updated pipeline on a feature branch in my git repository. When I'm done I want to clean them out of the history - they have no value other than to clutter things up.  I have to manually clean them one at a time in the user interface.

**Comment by jdholbrook81** [comment]
Need this feature. I have a classic build pipeline that ran on a weekly schedule for years. I'm unable to delete it with out manually deleting the runs through the UI due to error:

"One or more builds associated with the requested pipeline(s) are retained by a release. The pipeline(s) and builds will not be deleted."

The repo and release pipeline have already been deleted so nothing should be retaining the build.

**Comment by daniprado** [comment]
So, no news 2 years later?

**Comment by gautelo** [comment]
> So, no news 2 years later?

I doubt anything will happen unless it's taken to user voice and voted up there somehow, @daniprado. Another avenue is somebody new is employed who cares about this randomly and makes it happen. I feel like both of those are unlikely to happen. It's a pity, since it would probably not require a lot of effort to add this API.

**Comment by odelljl** [comment]
I would like to see some feedback on this issue as well.

**Comment by cmalard** [comment]
Meanwhile... some tool to help with it:

1. Retrieve the buildIds to delete:
  - either by using the [az cli](https://learn.microsoft.com/fr-fr/cli/azure/pipelines/build?view=azure-cli-latest#az-pipelines-build-list) if you want to filter by reason, result or status
  - or with this JS tool if you want to filter on the pipeline's name on your runs page (`https://dev.azure.com/<org>/<project>/_build?definitionId=<pipeline>`) through the web console:
```js
((filter) => Array.from(document.querySelectorAll('table tbody a'))
    .map(build => {
        const result = new RegExp('.*/_build/results\\?buildId=(\\d+)$').exec(build.getAttribute('href'));
        return result ? [result[1], build.textContent] : null;
    })
    .filter(build => build && build[1].includes(filter))
    .map(build => console.log(build.join(': ')) || build[0])
    .join(' ')
)('#test')
```
(replace `#test` with whatever filter you want ;) )
This logs the id with the run's name, allowing you to ensure you want to delete them:
<img width="544" alt="Image" src="https://github.com/user-attachments/assets/7a04dad5-bdda-441b-81a4-3c9594a33fdd" />

2. Use a sh script to do DELETE calls:
```sh
delete_ids() {
  local bearer='Bearer <token>'
  local url='https://dev.azure.com/<org>/<projectId>/_apis/build/builds/'

  local ids=("$@")
  local total=${#ids[@]}
  local count=0

  for id in "${ids[@]}"; do
    ((count++))
    echo -n "[$count/$total] Deleting ID $id:"
    curl -s -o /dev/null -w "%{http_code}\n" -X 'DELETE' "$url$id" \
      -X 'DELETE' \
      -H 'accept: application/json;api-version=6.0-preview.6;excludeUrls=true;enumsAsNumbers=true;msDateFormat=true;noArrayWrap=true' \
      -H "authorization: $bearer"
  done
}
delete_ids 314533 314530 314526 314522 314519 314516 314513
```
(replace `<token>`, `<org>` and `<projectId>` with your values, you can easily retrieve them by making a first deletion with your console network tab opened)
Save it in a file, for example `run.sh` in your user directory, then call it in your cli:
<img width="207" alt="Image" src="https://github.com/user-attachments/assets/e6276efe-683c-4ea2-b965-d1ea8e929d28" />

You may need to do this a few times, as Azure uses virtual scrolling on the webpage displaying the runs.
Enjoy 🍻
